devcareer/code-jammers-backend
# #175357378 Build CRUD API for tourist center

## What does this PR do ##
-  enables user to get one tourist center with specific id 
-  enables user to get all tourist centers
-  enables admin to add a tourist center
- enables admin to edit an existing tourist center in the database
- enables admin to delete a tourist center from the database

### Description of tasks to be completed? ###
- create get single tourist center route & controller
- create get multiple tourist centers route & controller
- create an add tourist center route & controller
- create a delete tourist center route & controller
- create an edit tourist center route & controller
- write tests for all tourist center routes & controllers

### How should this be manually tested/checked? ###
- Clone the PR
- Checkout to the branch `ft-user-should-get-touristcenters-175357378`
- Install the dependencies using `npm install`
- Create a `.env file` and define the following as follows:
   - `DEV_DATABASE_URL = <development database url>`
   - `TEST_DATABASE_URL = <test database url>`
   - `DATABASE_URL = <production database url>`
- run ```npm run migrate``` to create the database tables
- run ```npm run seed``` to seed the database
- Start the server by running ```npm run start-dev```
- Sign in via `http://localhost:3000/api/v1/users/signin` using a `POST` method with
```
          {
          "email: "ufuoma@gmail.com",
          "password": 12345
           }
```
and copy the token (without the quotes) from the response body.
#### Add Tourist Center ####
- In a new tab, under the `Authorization` tab, change the `TYPE` to `Bearer Token` and paste the token in the space available. Make a `POST` request to `http://localhost:3000/api/v1/admin/tourist-center/2e11e4a9-441b-4426-9521-39adc64ccfad` to add a tourist center. place the following in the request body and send to add the tourist center.
```
{
    "name": "Robben Island Museum",
    "location": "Cape Town",
    "gallery": "https://www.fodors.com/assets/destinations/45/robben-island-prison-robben-island-cape-town-south-africa_main.jpg",
    "about": "Robben Island is an island in Table Bay, 6.9 kilometres west of the coast of Bloubergstrand, north of Cape Town, South Africa. It takes its name from the Dutch word for seals, hence the Dutch/Afrikaans name Robbeneiland, which translates to Seal Island"
}

```
#### Get all Tourist Centers ####
- In a new tab, Make a `GET` request to `http://localhost:3000/api/v1/tourist-centers` to get all tourist centers
copy any tourist center id from the response data labeled as `id`
#### Get a Tourist Center ####
- In a new tab, Make a `GET` request to `http://localhost:3000/api/v1/tourist-center/<paste copied id>` to get the selected tourist center
#### UPDATE a Tourist Center ####
- In a new tab, Make a `PATCH` request to `http://localhost:3000/api/v1/admin/tourist-center/<paste copied id>` and place the fields to be updated in the request body.
    - request body sample:
```
{
    "countryId": "6003fb36-5112-463e-a1f9-c8944e72412f"
}
```
#### DELETE a Tourist Center ####
- In a new tab, Make a `DELETE` request to `http://localhost:3000/api/v1/admin/tourist-center/<paste copied id>` to delete the selected tourist center.

#### Tourist Center associations ####
-  make GET a request to url endpoint `http://localhost:3000/api/v1/countries`
it should return all the countries with their associated tourist centers in the database.


### Link to architecture
[architecture](https://docs.google.com/document/d/1PCD_1-7Vm0ohfEF65kngE7vsfx2RzJFDTpAuZcUYF78/edit?usp=sharing)

### What are the relevant pivotal tracker stories?
[ #175357378](https://www.pivotaltracker.com/story/show/175357378)
